qCC/ccCommandLineCommands.cpp

```diff
@@ -2124,36 +2133,172 @@ bool CommandApplyTransformation::process(ccCommandLineInterface& cmd)
 	{
 		return cmd.error(QObject::tr("No entity on which to apply the transformation! (be sure to open one with \"-%1 [filename]\" before \"-%2\")").arg(COMMAND_OPEN, COMMAND_APPLY_TRANSFORMATION));
 	}
-	
-	//apply transformation
-	if (!cmd.clouds().empty())
+
+	//create an entity vector
+	size_t nrOfClouds = cmd.clouds().size();
+	size_t nrOfMeshes = cmd.meshes().size();
+	size_t nrOfEntities = nrOfClouds + nrOfMeshes;
+	std::vector<std::pair<ccShiftedObject*, CLEntityDesc*>> entities;
+	entities.reserve(nrOfEntities);
+
+	//add clouds to the vector
+	for (CLCloudDesc& desc : cmd.clouds())
 	{
-		for (const CLCloudDesc& desc : cmd.clouds())
-		{
-			desc.pc->applyGLTransformation_recursive(&mat);
-			desc.pc->setName(desc.pc->getName() + ".transformed");
-		}
-		//save output
-		if (cmd.autoSaveMode() && !cmd.saveClouds("TRANSFORMED"))
+		entities.push_back({ desc.pc, &desc });
+	}
+
+	//add meshes to the vector
+	for (CLMeshDesc& desc : cmd.meshes())
+	{
+		bool isLocked = false;
+		ccShiftedObject* shifted = ccHObjectCaster::ToShifted(desc.mesh, &isLocked);
+		if (shifted && !isLocked)
 		{
-			return false;
+			entities.push_back({ shifted, &desc });
 		}
 	}
 
-	if (!cmd.meshes().empty())
+	//if the transformation is partly converted to global shift/scale
+	bool autoApplyPreviousGlobalShiftAndScale = false;
+	double previousScale = 1.0;
+	CCVector3d previousShift(0, 0, 0);
+
+	//process both clouds and meshes
+	for (size_t i = 0; i < nrOfEntities; i++)
 	{
-		for (const CLMeshDesc& desc : cmd.meshes())
+		CLEntityDesc& desc = *entities[i].second;
+		ccShiftedObject* shiftedEntity = entities[i].first;
+
+		ccGLMatrixd transMat(mat.data());
+		if (applyToGlobal)
+		{
+			// the user wants to apply the transformation to the global coordinates
+			CCVector3d globalShift = shiftedEntity->getGlobalShift();
+			double globalScale = shiftedEntity->getGlobalScale();
+
+			// we compute the impact to the local coordinate system without changing the
+			// actual Global Shift & Scale parameters (for now)
+			CCVector3d localTranslation = globalScale * (globalShift - transMat * globalShift + 2 * transMat.getTranslationAsVec3D());
+
+			// we switch to a local transformation matrix
+			transMat.setTranslation(localTranslation);
+
+			//test if the translated cloud coordinates were already "too large"
+			ccBBox localBBox = shiftedEntity->getOwnBB();
+			CCVector3d Pl = localBBox.minCorner();
+			double Dl = localBBox.getDiagNormd();
+			if (!ccGlobalShiftManager::NeedShift(Pl) && !ccGlobalShiftManager::NeedRescale(Dl))
+			{
+				//test if the translated (local) cloud coordinates are too large
```

Comment:
Cannot we skip this whole condition? Or at least create a -FORCE argument to pass through it?

If you say yes, i'll do it when i have opportunity probably next year.

Reply:
Yes why not

Reply:
skip or argument?:)

Reply:
Ah, so we can skip it, but in this case we should add a warning. Let me do it.

Reply:
it's all yours till january :)

Reply:
Ah but in this case the Global Shift will change in an uncontrolled way 😅 

Ok, so it's better to add an argument. I'll let you do it.

Reply:
and can you tell me why we check minCorner instead of maxCorner? probably we should check the highest absolute value, but usually that is the maxCorner since state grids never goes to negative territorry, i think.

Reply:
Well, it's just by consistency with the rest of the code, but it would make more sense in most cases indeed 👍

Reply:
ok, then happy new year. Probably i'm gonna be rush into real work tomorrow, we have a deadline on the first week. And i "wasted" (of course not)  my holidays with CC :) So probably i could take a look at this on the next half of the week.

Reply:
I've run some test. For me it is ready to merge